// @inngest/realtime/dist/subscribe/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSubscriptionToken = exports.subscribe = void 0;
const env_1 = require("../env");
const TokenSubscription_1 = require("./TokenSubscription");
/**
 * TODO
 */
const subscribe = async (
/**
 * TODO
 */
token, 
/**
 * TODO
 */
callback) => {
    const app = token.app;
    const api = app?.["inngestApi"];
    // Allow users to specify public env vars for the target URLs, but do not
    // allow this for signing keys, as they should never be on a client.
    const maybeApiBaseUrl = app?.apiBaseUrl ||
        (0, env_1.getEnvVar)("INNGEST_BASE_URL") ||
        (0, env_1.getEnvVar)("INNGEST_API_BASE_URL");
    const maybeSigningKey = api?.["signingKey"] || (0, env_1.getEnvVar)("INNGEST_SIGNING_KEY");
    const maybeSigningKeyFallback = api?.["signingKeyFallback"] || (0, env_1.getEnvVar)("INNGEST_SIGNING_KEY_FALLBACK");
    const subscription = new TokenSubscription_1.TokenSubscription(token, maybeApiBaseUrl, maybeSigningKey, maybeSigningKeyFallback);
    const retStream = subscription.getJsonStream();
    const callbackStream = subscription.getJsonStream();
    await subscription.connect();
    const extras = {
        getJsonStream: () => subscription.getJsonStream(),
        getEncodedStream: () => subscription.getEncodedStream(),
    };
    if (callback) {
        subscription.useCallback(callback, callbackStream);
    }
    else {
        callbackStream.cancel("Not needed");
    }
    return Object.assign(retStream, extras);
};
exports.subscribe = subscribe;
/**
 * TODO
 */
const getSubscriptionToken = async (
/**
 * TODO
 */
app, 
/**
 * TODO
 */
args) => {
    const channelId = typeof args.channel === "string" ? args.channel : args.channel.name;
    if (!channelId) {
        throw new Error("Channel ID is required to create a subscription token");
    }
    const key = await app["inngestApi"].getSubscriptionToken(channelId, args.topics);
    const token = {
        channel: channelId,
        topics: args.topics,
        key,
    };
    return token;
};
exports.getSubscriptionToken = getSubscriptionToken;
