// @inngest/realtime/dist/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.api = void 0;
const zod_1 = require("zod");
const env_1 = require("./env");
const util_1 = require("./util");
const tokenSchema = zod_1.z.object({ jwt: zod_1.z.string() });
exports.api = {
    async getSubscriptionToken({ channel, topics, signingKey, signingKeyFallback, apiBaseUrl, }) {
        let url;
        const path = "/v1/realtime/token";
        const inputBaseUrl = apiBaseUrl ||
            (0, env_1.getEnvVar)("INNGEST_BASE_URL") ||
            (0, env_1.getEnvVar)("INNGEST_API_BASE_URL");
        const devEnvVar = (0, env_1.getEnvVar)("INNGEST_DEV");
        if (inputBaseUrl) {
            url = new URL(path, inputBaseUrl);
        }
        else if (devEnvVar) {
            try {
                const devUrl = new URL(devEnvVar);
                url = new URL(path, devUrl);
            }
            catch {
                if ((0, util_1.parseAsBoolean)(devEnvVar)) {
                    url = new URL(path, "http://localhost:8288/");
                }
                else {
                    url = new URL(path, "https://api.inngest.com/");
                }
            }
        }
        else {
            url = new URL(path, (0, env_1.getEnvVar)("NODE_ENV") === "production"
                ? "https://api.inngest.com/"
                : "http://localhost:8288/");
        }
        const body = topics.map((topic) => ({
            channel,
            name: topic,
            kind: "run",
        }));
        const res = await (0, util_1.fetchWithAuthFallback)({
            authToken: signingKey,
            authTokenFallback: signingKeyFallback,
            fetch,
            url,
            options: {
                method: "POST",
                body: JSON.stringify(body),
                headers: {
                    "Content-Type": "application/json",
                },
            },
        });
        if (!res.ok) {
            throw new Error(`Failed to get subscription token: ${res.status} ${res.statusText} - ${await res.text()}`);
        }
        const data = await res.json();
        return tokenSchema.parse(data).jwt;
    },
};
