// @inngest/realtime/dist/channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.typeOnlyChannel = exports.channel = void 0;
const topic_1 = require("./topic");
/**
 * TODO
 */
const channel = (
/**
 * TODO
 */
id) => {
    // eslint-disable-next-line prefer-const, @typescript-eslint/no-explicit-any
    let channelDefinition;
    const topics = {};
    const builder = (...args) => {
        const finalId = typeof id === "string" ? id : id(...args);
        const topicsFns = Object.entries(topics).reduce((acc, [name, topic]) => {
            acc[name] = createTopicFn(finalId, topic);
            return acc;
        }, {});
        const channel = {
            name: finalId,
            topics,
            ...topicsFns,
        };
        return channel;
    };
    const extras = {
        topics,
        addTopic: (topic) => {
            topics[topic.name] = topic;
            return channelDefinition;
        },
    };
    channelDefinition = Object.assign(builder, extras);
    return channelDefinition;
};
exports.channel = channel;
/**
 * TODO
 */
const typeOnlyChannel = (
/**
 * TODO
 */
id) => {
    const blankChannel = {
        ...(0, exports.channel)(id),
        topics: new Proxy({}, {
            get: (target, prop) => {
                if (prop in target) {
                    return target[prop];
                }
                if (typeof prop === "string") {
                    return (0, topic_1.topic)(prop);
                }
            },
        }),
    };
    const ch = new Proxy(blankChannel, {
        get: (target, prop) => {
            if (prop in target) {
                return target[prop];
            }
            if (typeof prop === "string") {
                return createTopicFn(id, (0, topic_1.topic)(prop));
            }
        },
    });
    return ch;
};
exports.typeOnlyChannel = typeOnlyChannel;
const createTopicFn = (channelId, topic) => {
    return async (data) => {
        const schema = topic.getSchema();
        if (schema) {
            try {
                await schema["~standard"].validate(data);
            }
            catch (err) {
                console.error(`Failed schema validation for channel "${channelId}" topic "${topic.name}":`, err);
                throw new Error("Failed schema validation");
            }
        }
        return {
            channel: channelId,
            topic: topic.name,
            data,
        };
    };
};
