// @inngest/realtime/dist/middleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.realtimeMiddleware = void 0;
const inngest_1 = require("inngest");
const experimental_1 = require("inngest/experimental");
const realtimeMiddleware = () => {
    return new inngest_1.InngestMiddleware({
        name: "publish",
        init({ client }) {
            return {
                onFunctionRun({ ctx: { runId } }) {
                    return {
                        transformInput({ ctx: { step } }) {
                            const publish = async (input) => {
                                const { topic, channel, data } = await input;
                                const store = await (0, experimental_1.getAsyncCtx)();
                                if (!store) {
                                    throw new Error("No ALS found, but is required for running `publish()`");
                                }
                                const publishOpts = {
                                    topics: [topic],
                                    channel,
                                    runId,
                                };
                                const action = async () => {
                                    const result = await client["inngestApi"].publish(publishOpts, data);
                                    if (!result.ok) {
                                        throw new Error(`Failed to publish event: ${result.error?.error}`);
                                    }
                                };
                                return (store.executingStep
                                    ? action()
                                    : step.run(`publish:${publishOpts.channel}`, action)).then(() => {
                                    // Always return the data passed in to the `publish` call.
                                    return data;
                                });
                            };
                            return {
                                ctx: {
                                    /**
                                     * TODO
                                     */
                                    publish,
                                },
                            };
                        },
                    };
                },
            };
        },
    });
};
exports.realtimeMiddleware = realtimeMiddleware;
