// @inngest/realtime/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Realtime = void 0;
const zod_1 = require("zod");
var Realtime;
(function (Realtime) {
    Realtime.messageSchema = zod_1.z
        .object({
        channel: zod_1.z.string().optional(),
        topic: zod_1.z.string().optional(),
        data: zod_1.z.any(),
        run_id: zod_1.z.string().optional(),
        fn_id: zod_1.z.string().optional(),
        created_at: zod_1.z
            .string()
            .optional()
            .transform((v) => (v ? new Date(v) : undefined)),
        env_id: zod_1.z.string().optional(),
        stream_id: zod_1.z.string().optional(),
        kind: zod_1.z.enum([
            "step",
            "run",
            "data",
            "ping",
            "pong",
            "closing",
            "event",
            "sub",
            "unsub",
            "datastream-start",
            "datastream-end",
            "chunk",
        ]),
    })
        .transform(({ data, ...rest }) => {
        return {
            ...rest,
            data: data ?? undefined,
        };
    });
})(Realtime || (exports.Realtime = Realtime = {}));
